[Model][Misc] Add e5-mistral-7b-instruct and Embedding API

CatherineSue opened the pull request:
## Key Features of This Pull Request
This PR adds the e5-mistral-7b-instruct model and enables an E2E embedding vector generation. 

There are a few open PRs that add support for embedding models. Our PR **uniquely** addresses the following key issues:

### Integration of the OpenAI Server Front End
This PR includes comprehensive end-to-end (E2E) embedding functionality, spanning from the OpenAI Front End to the Back End.

### Turn off KV cache with embedding models
This PR introduces the capability to turn off KV cache when operating in embedding mode, which includes the block_tables and cache_engine.

## High Level Design
The embedding model can essentially be considered a special type of generative model with max_token=1 from an inference perspective. Both embedding and generative models (with max_token=1) require a single feedforward calculation without the need for generating any subsequent new tokens. The primary differences are:

The embedding model returns the hidden state, while the generative model takes an extra step to sample and return the first output token. In this sense, the embedding model is a subset of the generative model in terms of the calculations performed during the single feedforward operation on GPUs. They differ in their API specifications concerning request and output formats.

As a consequence, the serving of embedding models is:

- Independent of any technologies implemented for subsequent token-by-token generations; and
- They share the same technologies as generative models for prompt processing.

Therefore, in this PR, our current implementation focuses on bypassing vLLM’s components for subsequent token generations, such as **KV cache and CUDA Graph** to avoid unnecessary GPU memory, which in turn is useful to optimize the performance for embedding serving. If these technologies (e.g., CUDA Graph) are enhanced to improve prompt processing, they can be directly applied to embedding models. At a high level, our PR is **independent** of such crucial feature tensor parallelism, quantization, etc., which are addressed in **Milestone 1** below. We have tested it with tensor parallelism > 1 and it functions effectively. We are willing to work together to test/improve these crucial features where needed. 

## Benchmarking Results
On 1 H100 GPU, vLLM’s serving of the E5 embedding model reaches a max throughput up to 36.6k tokens/s and remains consistent across sequence lengths from 128 to 32k (when `gpu_memory_utilization` is 0.8).
On 1 A100 GPU, it reaches up to 15.3k tokens/second and  also remains consistent across sequence lengths. As a comparison, in a test of the latest ONNX, the speed is up to 12.6k tokens/sec when sequence length is low (256 tokens), is reduced to 8k tokens/sec when sequence length is 2k, and shows a general trend of getting worse when the sequent length is longer.
See the figure for more details of our testing results on H100 and A100
Note: The `throughput` is measured as the `"number of sequences in a batch" * "sequence length" / "end-to-end latency"`
![throughput_vs_batch_size_h100](https://github.com/CatherineSue/vllm/assets/8605658/d3e8525c-03a4-4a40-84c5-f5b9d767ae6c)

![throughput_vs_batch_size_a100](https://github.com/CatherineSue/vllm/assets/8605658/b8d28876-a6c0-44e3-aff0-1c3761ce3bb8)

## Design and Implementation Details
### Add Embedding generation

#### Add Embedding API to `entrypoints/openai`

- Add `EmbeddingRequest` and `EmbeddingResponse` to protocol.py
- Add `serving_embedding.py`
- Add `OpenAIServingEmbedding` to api_server
- Make `llm.py` work with embedding

#### Add Embedding in outputs and sequence
- Make an abstract base class of `RequestOutput` and `SequenceGroupOutput`.
- Add separate `Completion*Output` and `Embedding*Output`.
- Add `EmbeddingOutput`, `EmbeddingRequestOutput`, `RequestOutputFactory` and
`EmbeddingSequenceGroupOutput` to support processing the embedding output
sequence. 
- Update process output and sequence in *llm_engine to use embedding

#### Add MistralModel and embedding generation in llama.py

- Add `llama.embedding()` and `load_weights` in LlamaModel to support forward
and embedding vector generation
- Adapted from code examples in https://huggingface.co/intfloat/e5-mistral-7b-instruct
- Mistral uses LlamaModel.
- Use embedding when `embedding_mode` is True in model_runner
- Add `load_weights` in `llama.py` to support embedding models

### Disable KV Cache for Embedding serving

#### Skip `slot_mapping` with embedding mode

- `slot_mapping` is only used in model_executor/layers/attention.py when
kv_cache is not None. In embedding mode, we pass None kv_cache. So no
need to process slot_mapping

#### Turn off KV cache for embedding mode

The goal is to **disable** the block_table and cache_engine completely, so
we don't consider allocating blocks for KV cache for embedding mode
- Add `embedding_mode` to `ModelConfig` and `SchedulerConfig`
- Add a `BlockSpaceManagerProxy `to control the block management behavior 
for embedding mode

#### Update parameters for max batching
- Add `profile_max_batched_tokens_for_embedding` to profile the
max_num_batched_tokens for embedding server mode
- Return max_batch_size as each Ray worker runs the profiling once
- Enable embedding profiling in `ray_gpu_executor.py` to support tensor parallelism > 1

## Notes
### Overlaps with other PRs 
This PR overlaps with the following [Milestones](https://github.com/vllm-project/vllm/pull/3187#issuecomment-2016902999) in [Supporting embedding models](https://github.com/vllm-project/vllm/pull/3187#top) #3187
### Milestone 1: Basic functionality using Hugging Face models [Partially completed]
Note: Instead of using `LLM.encode()`, this PR currently adds `embedding()` to `LlamaModel`, and then keeps `LLM.generate()` for embedding.

#### G) Update parameters for max batching [Completed]
We introduced `profile_max_batched_tokens_for_embedding` in `gpu_executor.py` to support the maximum number of tokens the GPU can take in one batch. 

### Milestone 2: Implement the model properly [Completed]
This PR focuses on adding the e5-mistral-7b-instruct model, which can utilize llama.py. So it already uses the vLLM layer primitives.

### Milestone 4: Add the OpenAI Server Front End [Completed]
This PR has implemented the Embedding API to `entrypoints/openai`. 

### Discrepancies with other discussions
This PR didn't implement the following in this [discussion](https://github.com/vllm-project/vllm/pull/3187#issuecomment-2016937039) in [Supporting embedding models](https://github.com/vllm-project/vllm/pull/3187#top) #3187
##### Move finish sequence logic to check_stop
Currently, the logic is in `llm_engine._process_model_output()`
The logic should be in `llm_engine._check_stop`
##### Automatically detect that the model is an embedding model
The user should not have to specify that it is an embedding model
Somewhere in the vllm code, create a registry that selects which models are embedding and which models are decoders
#### F) Update Pass `EmbeddingParams` around instead of `SamplingParams`
note: this is going to require a lot of work
**Note: This PR passes `SamplingParams()` to the LLMEngine and disabled the use of it in embedding mode.** As separating `EmbeddingParams` and `SamplingParams` requires changes to the UX, it would be easier to discuss and review in a following PR.

### Discussion Points and Considerations
#### Simplifying the Workflow
- Evaluate the possibility of consolidating embedding-related workflows into existing structures with minimal branching logic. This could involve using if-else statements (specifically in `model_runner.py`) or integrating embedding as a subset of generation needs. 
#### Soundness of Profiling 
- Evaluate if the current `profile_max_batched_tokens_for_embedding` is sufficient to support the maximum number of tokens in one batch without causing CUDA OOM.

CatherineSue committed:
Disable KV Cache for Embedding serving and Add Embedding generation

Apply suggestions from code review

- Fix typo EmeddingResponseData
- Apply nit suggestion on EmbeddingRequest check

Co-authored-by: Roger Wang <136131678+ywang96@users.noreply.github.com>

Change v1/embedding to v1/embeddings

Add check for EmbeddingRequest input tokens num length

Add TODO comments for vllm-specific ValidationError in embedding

Add miscellaneous small fixes

- Format files
- Remove duplicate imports in llm_engine
- Fix imports in llama_embedding
- Remove doc in arg_utils

Separate llama_embedding from llama

- Add LlamaEmbeddingModel in llama_embedding.py
- Add _EMBEDDING_MODELS in models/__init__
- Move _MODELS to _GENERATION_MODELS

Refactor ModelRunner and remove profiling for embedding

Refactor ModelRunner
- Add EmbeddingModelRunner to override execute_model

Remove profiling for embedding
- Remove profile_max_batched_tokens_for_embedding in worker
- Set default max_num_batched_tokens to 32768 for embedding

Add BlockSpaceManagerV3 and remove DummyBlockSpaceManager

- Use BlockSpaceManagerV3 for simple block management for embedding
- Fix Scheduler.prompt_limit for embedding
- Fix typo

Replace 0.8 in profiling max_num_batched_tokens

- Utilize gpu_memory_utilization factor instead

Refactor RequestOutput and SequenceGroupOutput

- Add an abstract base class of RequestOutput, separate Completion and
Embedding
- Add an abstract base class of SequenceGroupOutput, separate
Completion and Embedding

Add docstring for embedding_mode in ModelConfig

Remove max_tokens in EmbeddingRequest

Fix generation in model_runner.py and formatting
- Fix model.sample and embedding generation in model_runner.py
- Format files changed
- Fix embedding in llama.py as input_tokens' shape has changed
- Enable embedding in ray_gpu_executor.py

Add Embedding generation

Add Embedding API to entrypoints/openai

- Add EmbeddingRequest and Response to protocol
- Add serving_embedding
- Add OpenAIServingEmbedding to api_server
- Make llm.py work with embedding

Add Embedding in outputs and sequence

- Add EmbeddingOutput, EmbeddingRequestOutput, RequestOutputFactory and
EmbeddingSequenceGroupOutput to support process the embedding output
sequence
- Update process output and sequence in *llm_engine to use embedding

Add MistralModel and embedding generation

- Add embedding and load_weights in LlamaModel to support forward
- Adapted from code examples in https://huggingface.co/intfloat/e5-mistral-7b-instruct
- Ensures the ops are run on GPU.
- Mistral uses LlamaModel
- Use embedding when embedding_mode is True in model_runner

Uniform typing in multiple files

- Add EmbeddingRequestOutput to the output typing in llm,
async_llm_engine
- Add validation for max_tokens in EmbeddingRequest
- Fix __repr__ of EmbeddingSequenceGroupOutput

Disable KV Cache for Embedding serving

Update profiling max_batch_size logic
- Return max_batch_size as each Ray worker runs the profiling once

Skip slot_mapping with embedding mode
slot_mapping is only used in model_executor/layers/attention.py when
kv_cache is not None. In embedding mode we pass None kv_cache. So no
need to process slot_mapping.

Disable KV cache for embedding mode
- Add embedding_mode to ModelConfig and SchedulerConfig
- Add profile_max_batched_tokens_for_embedding to profile the
max_num_batched_tokens for embedding server mode
- Add DummyBlockManager for No-op block management in embedding mode
- Add load_weights in llama.py to support embedding models

rkooo567 reviewed (commented):
Please ignore. wrong comment

cadedaniel commented on vllm/core/block_manager.py:
```diff
@@ -592,3 +593,62 @@ def mark_blocks_as_computed(self, seq_group: SequenceGroup):
+                enable_caching=cache_config.enable_prefix_caching,
+            )
+
+    def __getattr__(self, item):
+        if self._embedding_mode:
+            # Handle methods that need to return True or any specific value
+            if item in ["can_swap_in", "can_swap_out"]:
+                return lambda *args, **kwargs: True
+            if item in ["can_allocate"]:
+                return lambda *args, **kwargs: AllocStatus.OK
+            if item in ["get_num_free_gpu_blocks", "get_num_free_cpu_blocks"]:
+                return lambda *args, **kwargs: 1
+            # For other methods, return a no-op or a harmless default
+            return lambda *args, **kwargs: None
+        # In the generation mode, just proxy through to the real manager
+        return getattr(self._real_block_manager, item)
```
We should implement the new interface defined here instead of getattr proxy https://github.com/vllm-project/vllm/blob/main/vllm/core/interfaces.py#L22

it'll need some tests too!

CatherineSue replied:
Done. Added a new impletation.

cadedaniel commented on vllm/executor/gpu_executor.py:
```diff
@@ -37,7 +37,10 @@ def __init__(
 
         # Profile the memory usage and initialize the cache.
-        self._init_cache()
+        if not self.model_config.embedding_mode:
+            self._init_cache()
+        else:
+            self._init_embedding()
```
can we find a common abstraction for both types?

CatherineSue replied:
Removed `_init_embedding`

cadedaniel commented on vllm/engine/arg_utils.py:
```diff
@@ -209,7 +210,10 @@ def add_cli_args(
             'the model executor, which can range from 0 to 1.'
-            'If unspecified, will use the default value of 0.9.')
+            'If unspecified, will use the default value of 0.9.'
+            'Recommend to set it to 0.8 for embedding models '
+            'to effectively handle memory demands while minimizing '
+            'the risk of OOM errors.')
```
I feel this should go into documentation, not CLI help (it is specific only to embedding models)

CatherineSue replied:
Removed the documentation. Embedding won't use this parameter.

cadedaniel commented on vllm/core/scheduler.py:
```diff
@@ -83,18 +83,19 @@ def __init__(
 
-        self.prompt_limit = min(self.scheduler_config.max_model_len,
-                                self.scheduler_config.max_num_batched_tokens)
+        if self.scheduler_config.embedding_mode:
+            # Profile this value if necessary
+            self.prompt_limit = self.scheduler_config.max_num_batched_tokens
+        else:
+            self.prompt_limit = min(
+                self.scheduler_config.max_model_len,
+                self.scheduler_config.max_num_batched_tokens)
```
is it possible to use the same abstraction for both model types? trying to reduce branches in the codebase

CatherineSue replied:
Removed this.

CatherineSue commented:
An overview of tasks derived from the pull request discussions:

### Immediate changes
core/
- [x] Make a `BlockSpaceManagerV3`
- [x] Remove `self.prompt_limit` if branch in scheduler.py

worker/ and executor/
- [x] Remove `profile_max_num_batched_tokens` for embedding
- [x] Evaluate to set a hardcoded `batch_size` or `max_num_batched_tokens` (32k as Woosuk and Zhuohan suggested) instead of profiling

models/
- [x] Make a new embedding_models_dict
- [x] Make a llama_embedding.py
- [x] Check how HuggingFace(HF) generates embedding

Separate model_runner and embedding_model_runner
- [x] Implement `execute_model` separately in model_runner and embedding_model_runner

### Further evaluation
- Evaluating the possibility of using `config.json` to distinguish whether a model is for embedding or generation, e.g. `XForCausalLM` as a generation model. Check edge cases for fine-tuning.
- Separate LLM and LLMEmbedding
- Separate SamplingParams and EmbeddingParams
- Evaluate latency on individual request level

### HF embeddings
HF's [`sentence_transformers`](https://huggingface.co/sentence-transformers) provide sentences, texts, and image embeddings. They use an [`encode()`](https://github.com/UKPLab/sentence-transformers/blob/fa72a30ad0419b721439985e4d291935b79fd6f3/sentence_transformers/SentenceTransformer.py#L249) for computing sentence embeddings. And [`Pooling` class](https://github.com/UKPLab/sentence-transformers/blob/fa72a30ad0419b721439985e4d291935b79fd6f3/sentence_transformers/models/Pooling.py#L68-L73) supports different types of pooling, including `lasttoken`, which `e5-mistral-7b-instruct` uses. 

Adhering to the same design would require further discussion and evaluation of the points mentioned above.

ichernev commented on vllm/entrypoints/openai/serving_embedding.py:
```diff
@@ -0,0 +1,150 @@
+        async for i, res in result_generator:
+            if await raw_request.is_disconnected():
+                # Abort the request if the client disconnects.
+                await self.engine.abort(f"{request_id}-{i}")
```
Shouldn't we abort all outstanding requests -- i.e the ones for which `final_res_batch[i]` is `None`

CatherineSue replied:
I applied the same logic here as  [serving_completion](https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/openai/serving_completion.py#L184-L191). Can you explain the reason to abort all outstanding requests here?

ichernev replied:
If I'm understanding the code correctly, if the `raw_request` is disconnected, i.e the client disconnects, all sequences in the client request are effectively aborted. The existing code aborts the currently looked-at request, but then returns, which means all the other un-finished requests are not aborted, which is not ideal. We might as well not abort this one.

If this code is duplicated, it might be OK to fix later everywhere :)

ichernev commented on vllm/entrypoints/openai/serving_embedding.py:
```diff
@@ -0,0 +1,150 @@
+    ]
+
+    async def consumer():
+        while not all(finished) or not queue.empty():
```
The only use of `finished` is to see if they're all finished. Isn't a simple counter simpler (and faster)?

CatherineSue replied:
I applied the same logic as [serving_completion](https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/openai/serving_completion.py#L77-L83). I don't have a strong opinion on the simplicity. A counter might be slightly faster.

ichernev replied:
I understand. I guess another PR for all affected code might be better. I was just reviewing this code and noticed the (slight) inefficiency.

ichernev commented on vllm/entrypoints/openai/serving_engine.py:
```diff
@@ -175,6 +177,10 @@ def _validate_prompt_and_tokenize(
             prompt).input_ids
         token_num = len(input_ids)
 
+        # Note: EmbeddingRequest doesn't have max_tokens
```
I disagree. There is a property in config called `max_position_embeddings`:
- https://huggingface.co/intfloat/e5-mistral-7b-instruct/blob/main/config.json#L12
- https://huggingface.co/intfloat/e5-base-v2/blob/main/config.json#L15

And it's normally something rather small. `e5-mixtral-7b-instruct` is a bit unique with max context 32k :)

CatherineSue replied:
The [OpenAI official API](https://platform.openai.com/docs/api-reference/embeddings) doesn't include max_tokens in the embedding request.

ichernev replied:
So that means any old (embeddings) model can compute embeddings on any length of input without issues?

CatherineSue replied:
No, it still can't exceed the model's max context length, which the scheduler checks.

CatherineSue replied:
vLLM checks both fields in model_config: https://github.com/vllm-project/vllm/blob/main/vllm/config.py#L932
So added check for `token_num > self.max_model_len` in `serving_engine`.

ichernev replied:
Yeah, I guess I didn't understand the comment about `max_tokens`. I didn't realize this function also works for non-embedding requests.

ichernev commented on vllm/entrypoints/openai/serving_embedding.py:
```diff
@@ -0,0 +1,150 @@
+                                         SamplingParams(),
+                                         f"{request_id}-{i}",
+                                         prompt_token_ids=input_ids))
+        except ValueError as e:
```
If this exception is thrown after some generators are created, shouldn't we cancel them? Or perhaps first prep the input_ids for all sequences, and then just append the generators (which shouldn't throw).

CatherineSue replied:
Same as above. The logic here is the same as [serving_completion](https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/openai/serving_completion.py#L151-L159). 
The requests with exceptions thrown here are not thrown by appending the generators, right?

ichernev replied:
Well, by looking at the code the append line appends an un-awaited promise. The promise, if not awaited is not even executed (the code that fulfills it, that is), so it shouldn't be able to throw anything. The other lines however throw stuff, if there is an issue with the input.

Thinking about this -- if an exception is throw the generators are not awaited, so they shouldn't actually deploy any work to the engine, so I guess that part of the code is fine.

pzhao1799 reviewed (commented):
Preliminary review, will continue to add comments as I continue reading through this PR. Thanks a lot for all the hard work so far!!!

ywang96 reviewed (commented):
Thank you very much for your contribution @CatherineSue!

I took a pass and left some comments and suggestions. Overall the design and implementation is pretty straightforward to me, and thank you for creating the no op blockmanager to separate the logic.


Some thoughts I would like to throw out there:
- The current implementation doesn't allow an engine to do generation and embedding at the same time (since the `embedding_mode` is passed to initialize the engine), so I wonder if it's actually worth the effort to create a separate `EmbeddingEngine` since a lot of logics in `LLMEngine` are completely not needed, and this separation could also make a lot of higher level APIs cleaner. 
- If the end goal is indeed to have engine to support both at the same time, then we should think about a clean way to support it, as well as supporting other vanilla embedding models.

Happy to discuss and hear your thoughts  on this!

CatherineSue commented on vllm/core/block_manager_v3.py:
```diff
@@ -0,0 +1,84 @@
+from vllm.sequence import Sequence, SequenceGroup
+
+
+class BlockSpaceManagerV3(BlockSpaceManager):
+    """A simple version of BlockSpaceManager for use in environments
+    where block management is not required.
+
+    This class provides the same interface as BlockSpaceManager, but its
+    methods perform no actions or return simple values like True in specific
+    actions. It's designed to be used in scenarios where the overhead of
+    block management is unnecessary, such as in an embedding environment.
+    """
```
I agree. I had a quite debate on the naming. Is DummyBlockSpaceManager better or SimpleBlockSpaceManager? A few functions here always return true such as `can_allocate`, and `can_swap_out`. So it feels not that dummy. 
Additional thought: If there is plan to make Scheduler more extensible, maybe in the future we can simplify the logic of schedule with embedding, thus further cleanse this class as well.

CatherineSue commented on vllm/entrypoints/openai/protocol.py:
```diff
@@ -381,6 +389,21 @@ class CompletionStreamResponse(BaseModel):
     usage: Optional[UsageInfo] = Field(default=None)
 
 
+class EmeddingResponseData(BaseModel):
```
Thanks for noticing this!

CatherineSue commented on vllm/model_executor/models/llama_embedding.py:
```diff
@@ -0,0 +1,327 @@
+        # Calculate the flat indices for the last token of each sequence
+        last_token_flat_indices = cum_lengths - 1
+
+        # Ensure last_token_flat_indices is compatible for direct indexing
```
IMO it is ok without assertion here. I added detailed documentation here since there is a change in input_tokens (vLLM removed padding) and I changed the implementation here.

CatherineSue committed:
Make LlamaEmbeddingModel generate normalized embeddings

Add _check_embedding_mode to OpenAIServingEmbedding

CatherineSue commented on vllm/entrypoints/openai/serving_embedding.py:
```diff
@@ -0,0 +1,152 @@
+    return consumer()
+
+
+class OpenAIServingEmbedding(OpenAIServing):
+
+    def __init__(self, engine: AsyncLLMEngine, served_model: str):
+        super().__init__(engine=engine,
+                         served_model=served_model,
+                         lora_modules=None)
```
Added a check

CatherineSue commented:
@ywang96 thanks for the initial review!

I have applied the suggestions and resolved some comments.

> The current implementation doesn't allow an engine to do generation and embedding at the same time (since the embedding_mode is passed to initialize the engine), so I wonder if it's actually worth the effort to create a separate EmbeddingEngine since a lot of logics in LLMEngine are completely not needed, and this separation could also make a lot of higher level APIs cleaner.

I agree that having a separate EmbeddingEngine has a lot of benefits. Besides making the design cleaner, adding more functionalities in embedding and supporting more embedding models would be easier. Happy to discuss detailed design. It might be good for a following PR as the current is big.

CatherineSue commented:
@ywang96 A question related to design:

[intfloat/e5-mistral-7b-instruct](https://huggingface.co/intfloat/e5-mistral-7b-instruct) requires an eos_token_id appended at the end each input. See:
```
# Tokenize the input texts
batch_dict = tokenizer(input_texts, max_length=max_length - 1, return_attention_mask=False, padding=False, truncation=True)
# append eos_token_id to every input_ids
batch_dict['input_ids'] = [input_ids + [tokenizer.eos_token_id] for input_ids in batch_dict['input_ids']]
batch_dict = tokenizer.pad(batch_dict, padding=True, return_attention_mask=True, return_tensors='pt')

```

I checked a few embedding models, it seems specific to this model. Since input_ids are processed in [ModelRunner._prepare_prompt](https://github.com/vllm-project/vllm/blob/main/vllm/worker/model_runner.py#L223), is there a good place to inject eos_token_id to the end of each input?

robertgshaw2-redhat reviewed (changes requested):
Thanks for the great work on this.

@ywang96 and I are syncing this morning on this topic.

Will revert with detailed plan to get this merged

CatherineSue commented on vllm/config.py:
```diff
@@ -559,19 +564,24 @@ def __init__(
+        # for higher throughput.
+        # Otherwise, use the greater of max_model_len or 2048 as a default for
+        # higher throughput.
+        self.max_num_batched_tokens = (
```
@robertgshaw2-neuralmagic we had the profiling before if you look at the [initial commit](https://github.com/vllm-project/vllm/pull/3734/commits/c34cbb9f7fb101adcda4de2f88b78c8e7442f856#diff-53bf7aa55319c7c5f0213caca6122be63791a43ff3a92df607b2ff22e3cd1922). However, we decided to remove it after discussing with a few maintainers. Maximum the number of batched tokens to a higher number doesn't help to increase the performance.

CatherineSue commented on vllm/model_executor/models/llama_embedding.py:
```diff
@@ -0,0 +1,328 @@
+            config.hidden_size,
+            org_num_embeddings=config.vocab_size,
+        )
+        self.layers = nn.ModuleList([
```
We discussed this with a few maintainers in a previous video chat. The common feedback was that to separate llama embedding from llama. If adding LlamaEmbeddingModel in llama.py is ok I will change that

robertgshaw2-neuralmagic replied:
Okay, we can have a separate file, but should import `LlamaModel` from `llama.py` rather than re-writing the model I think

CatherineSue replied:
Done

CatherineSue commented on vllm/executor/gpu_executor.py:
```diff
@@ -42,7 +42,8 @@ def __init__(
 
         # Profile the memory usage and initialize the cache.
-        self._init_cache()
+        if not self.model_config.embedding_mode:
```
We haven't tested this in cpu_executor or neuron_executor yet.

CatherineSue commented on vllm/engine/arg_utils.py:
```diff
@@ -390,20 +391,21 @@ def add_cli_args(
             default=None,
             help='The number of speculative tokens to sample from '
             'the draft model in speculative decoding')
+        parser.add_argument("--embedding-mode",
```
This idea was proposed before. The idea is that xxxForCausalLM is usually for generation. But we need to investigate if there will be edge cases that use xxxModel for fine tuning instead of embedding.

robertgshaw2-neuralmagic replied:
HF has the following standards:
- `XXXModelForCausalLM` <==>  decoder with LMHead
- `XXXModelForConditionalLM` <==> encoder-decoder with LMHead
- `XXXModel` <==> embedding

So, we are good to go with this assumption. Plus we already have a registry of explicit models we support so I think this is okay

CatherineSue replied:
Done

CatherineSue replied:
Yup. It got mixed up during rebasing.

robertgshaw2-neuralmagic commented:
cc @simon-mo 

@CatherineSue 

@ywang96 and I discussed.  

There are a few things we want to do for the final implementation, but we want to take an approach of landing something close to the current version and then following up with incremental work.

## Needed For Merge

### Tests

This is a big feature that needs end-to-end testing. I would suggest that we focus in the following areas:

- `Model correctness`: compare vs sentence-transformers or hugging face implementation of the model. Use L2 norm of the differences between the embeddings. This can use the LLM engine.

- `Server API correctness`: show that querying the API gets what is expected from the client side. Check out the existing tests for the OpenAI API server for inspiration as to how to write these tests

### Clean up `llama_embedding.py`

We currently re-implement all the `llama` layers.  I am okay with having a separate file for `llama_embedding.py`, but we should import `LlamaModel` and use that rather than re-writing all the layers of the model. For example:

```python3
from vllm.model_executor.models.llama import LlamaModel

class LlamaEmbeddingModel(nn.Module):
    def __init__():  
         self.model = LlamaModel(**)
         
    def forward(inputs):
        return self.model(inputs)

    # currently called embedding
    def pool(hidden_states):
       # same as current embedding function
```

### Update Terminology from `embedding` to `pool`

This is to be consistent with HF / sentence-transformers, which use this terminology for translating between the final hidden states and output embedding.

So specifically
- `LlamaEmbeddingModel` should use `pool` instead of `embedding`
- `EmbeddingModelRunner

### Update `LLM.generate` API to log a warning if used with the embedding model
- Make a note that this interface is experimental, and that `LLM.encode` will replace it soon


### Add Incompatible Feature Guards

The following features are incompatible with embedding models:
- Spec Decode
- Chunked Prefill
- Automatic Prefix Caching
- Neuron / CPU
- Fp8 KV cache

If these features are specified for an embedding model, we should either fail or log a warning. I would be okay with doing this in a follow up PR if it is not straightforward.


I will make a follow up note for

robertgshaw2-neuralmagic commented:
## Follow Ups Post Merge

After we merge this initial implementation, we can refactor in the following way:

#### Replace `SamplerXXX` with `PoolerXXX`
- Swap `SamplingParameters` for `PoolerParameters`, which only had the data needed for embedding models
- Swap `SamplerOutput` for `PoolerOutput`, which only has the data needed for embedding models
- Pipe all this info around the various layers of the engine

#### `LLM.encode`

Deprecate embedding models from `LLM.generate()`. 

Instead expose `LLM.encode()`, which accepts `PoolerParams` and returns `EmbeddingRequestOutput`


#### Create Generic `Pooler` 

Create a generic `Pooler()` that corresponds to `Sampler()` (currently, we implement pooling logic in `LlamaEmbeddingModel.embedding()` rather than in a shared class for each . Pooler could be instantiated with [sentence transformer config](https://huggingface.co/BAAI/bge-m3/blob/main/1_Pooling/config.json)

This will allow us to support more complex methods like ColBERT, sparse, etc over time.

#### Refactor `LLMEngine`
- Currently `_process_sequence_group_outputs` is very complicated and focused on CompletionRequests
- Add abstract called `SequenceGroupProcessor`, with subclasses `SequenceGroupProcessorEmbedding` `SequenceGroupProcessorCompletion`. Each of these is responsible for implementing `_process_sequence_group_outputs` using `PoolerOutput` or `SamplerOutput`   ##respectively

CatherineSue commented:
@robertgshaw2-neuralmagic thanks for the feedback. Working on resolving the comments and getting the checklist. Will update it soon.

robertgshaw2-neuralmagic commented:
> @robertgshaw2-neuralmagic thanks for the feedback. Working on resolving the comments and getting the checklist. Will update it soon.

Thanks @CatherineSue 

Apologies for the delay on getting this reviewed and thank you so much for your contribution :)

CatherineSue committed:
Rename BlockSpaceManagerV3 to EmbeddingModelBlockSpaceManager

CatherineSue committed:
Clean up LlamaEmbeddingModel

- Add Pooler, PoolerOutput
- Reuse LlamaModel for LlamaEmbeddingModel
- Add PoolingParams to SequenceGroup

CatherineSue committed:
Use ModelRegistry to enable ModelConfig.embedding_mode

- Remove embedding_mode in engine args
- Use re to search for XXXModel to match embedding models

CatherineSue committed:
Separating PoolerOutput, PoolingParams from SamplingXXX

- Update SequenceGroup initialization in tests/
- Use PoolingParams and PoolerOutput for embedding_model_runner
- Add PoolingMetadata

CatherineSue committed:
Separating LLM.encode() from LLM.generate()

- Add LLM.encode()
- Add encode() in AsyncLLMEngine

CatherineSue committed:
Add tests for LlamaEmbeddingModel and OpenaiAPI server embedding

CatherineSue commented on vllm/model_executor/models/llama_embedding.py:
```diff
@@ -0,0 +1,328 @@
+        seq_outputs = []
+        for output in outputs_normalized:
+            seq_outputs.append(
+                EmbeddingSequenceGroupOutput(embeddings=output.tolist()))
```
Moved it to Pooler

CatherineSue commented:
@robertgshaw2-neuralmagic I resolved all the comments. Here's an overview of the tasks in the new commits:

- [x]  Rename `BlockSpaceManagerV3` to `EmbeddingModelBlockSpaceManager`
- [x]  Use ModelRegistry to check for embedding models
- [x]  Tests
- [x]  Clean up `llama_embedding.py`
- [x]  Update terminology from `LlamaEmbeddingModel.embedding` to `LlamaEmbeddingModel.pooler`
- [x]  Replace `SamplerXXX` with `PoolerXXX`
I added `Pooler`, `PoolingParams`, `PoolerOutput`, and `PoolingMetadata`. Note that `Pooler` is not following sentence_transformer's config. I didn't have time to finish it.
- [x] `LLM.encode`
I have separated it from `LLM.generate`

- [x] Log warning from LLM.generate() that this API will change for embedding models
Since I have separated it so I didn't add warning.

robertgshaw2-neuralmagic commented:
@CatherineSue - awesome! 

Are you planning to resolve the merge conflicts? 

Should I review now?

CatherineSue commented:
@robertgshaw2-neuralmagic 
I can resolve them if it is easier for you to review. Might take a while, ETA tonight. Does it work for you?

robertgshaw2-neuralmagic commented:
that works, ill review tomorrow

robertgshaw2-neuralmagic commented:
ping me when ready

ywang96 commented:
I can take a first pass too whenever it's ready if @robertgshaw2-neuralmagic doesn't get there before me :)

CatherineSue committed:
Fix errors caused by rebase

- Update execute_model and prepare_input_tensors in
embedding_model_runner.py to follow model_runner.py
= Remove unused imports

CatherineSue commented:
@robertgshaw2-neuralmagic @ywang96 Just finished rebase

ywang96 commented on vllm/worker/embedding_model_runner.py:
```diff
@@ -0,0 +1,126 @@
+        seq_group_metadata_list: Optional[List[SequenceGroupMetadata]],
+        kv_caches: List[torch.Tensor],
+    ) -> Optional[PoolerOutput]:
+        (input_tokens, input_positions, attn_metadata, sampling_metadata,
```
```suggestion
        (input_tokens, input_positions, attn_metadata, pooling_metadata,
```

ywang96 commented on vllm/worker/embedding_model_runner.py:
```diff
@@ -0,0 +1,126 @@
+            execute_model_kwargs.update({"image_input": multi_modal_input})
+        hidden_states = model_executable(**execute_model_kwargs)
+
+        return self.model.pooler(hidden_states=hidden_states,
+                                 attention_metadata=attn_metadata)
```
I'm a bit confused and might have missed something here: Shouldn't we always pass `pooling_metadata` to `pooler`, even if it's not used by the pooler for some models? (Otherwise `pooling_metadata` is not used anywhere)

ywang96 replied:
> To use it instead would involve CPU <-> GPU ops.

Yep I figured that's the reason why you're using attention metadata to retrieve prompt lens. It just seems weird to me from the developer perspective that we have `pooling_metadata` but it's not used by pooler or anywhere else.

CatherineSue replied:
Added `PoolingTensor` to create the tensor from `PoolingMetadata`

ywang96 commented on vllm/engine/async_llm_engine.py:
```diff
@@ -592,8 +600,8 @@ async def generate(
         Yields:
-            The output `RequestOutput` objects from the LLMEngine for the
-            request.
+            The output `RequestOutput` or `EmbeddingRequestOutput` objects
+            from the LLMEngine for the request.
```
```suggestion
            The output `CompletionRequestOutput` objects from the LLMEngine
            for the request.
```

ywang96 commented on vllm/engine/async_llm_engine.py:
```diff
@@ -638,25 +646,128 @@ async def generate(
+            multi_modal_data: Multi modal data per request.
+
+        Yields:
+            The output `RequestOutput` or `EmbeddingRequestOutput` objects
+            from the LLMEngine for the request.
```
```suggestion
            The output `EmbeddingRequestOutput` objects from the LLMEngine 
            for the request.
```

ywang96 commented on vllm/engine/async_llm_engine.py:
```diff
@@ -638,25 +646,128 @@ async def generate(
+        lora_request: Optional[LoRARequest] = None,
+        multi_modal_data: Optional[MultiModalData] = None
+    ) -> AsyncIterator[EmbeddingRequestOutput]:
+        """Generate outputs for a request.
```
```suggestion
        """Generate outputs for a request from an embedding model.
```

ywang96 commented on vllm/engine/async_llm_engine.py:
```diff
@@ -638,25 +646,128 @@ async def generate(
+        prompt: Optional[str],
+        params: Union[SamplingParams, PoolingParams],
+        prompt_token_ids: Optional[List[int]] = None,
+        arrival_time: Optional[float] = None,
```
it looks like neither `.generate` nor `.encode` sets `arrival_time` before calling `process_request`, do we need it in the function signature then?

CatherineSue replied:
Removed `arrival_time` in function signature

ywang96 reviewed (commented):
Thank you @CatherineSue for addressing the plan from @robertgshaw2-neuralmagic! I left a few questions, comments and suggestions, and will leave Robert to do a final review.

CatherineSue commented on vllm/worker/embedding_model_runner.py:
```diff
@@ -0,0 +1,126 @@
+                                          skip_sampling=True)
+
+        # Prepare PoolingMetadata
+        pooling_metadata = self._prepare_pooling(seq_group_metadata_list,
+                                                 sampling_metadata.prompt_lens)
```
Thanks for catching this, My oversight.

CatherineSue replied:
Applied the change

CatherineSue commented:
@ywang96 thanks for the quick review. Will start addressing comments after Robert has another look.

robertgshaw2-redhat committed:
Update vllm/engine/async_llm_engine.py

Co-authored-by: Roger Wang <136131678+ywang96@users.noreply.github.com>

robertgshaw2-neuralmagic commented on vllm/worker/worker.py:
```diff
@@ -83,7 +86,8 @@ def __init__(
         self.cache_engine: CacheEngine
-        self.gpu_cache: List[torch.Tensor]
+        # Initialize gpu_cache as embedding models don't initialize kv_caches
+        self.gpu_cache: List[torch.Tensor] = None
```
Should be

`Optional[List[torch.tensor]]`

CatherineSue replied:
Done.

robertgshaw2-neuralmagic commented on vllm/worker/embedding_model_runner.py:
```diff
@@ -0,0 +1,126 @@
+            lora_requests,
+            lora_mapping,
+            multi_modal_input,
+        ) = super().prepare_input_tensors(seq_group_metadata_list,
```
Can't `skip_sampling` be infered from the `SeqGroupMetadataList`?

CatherineSue replied:
Can you elaborate how? We would need to check `if seq_group_meta_data.sampling_params is not None for seq_group_meta_data in SeqGroupMetadataList`?

CatherineSue replied:
It seems there is some logic change in `ModelRunner.prepare_input_tensors()`, especially related to SamplingMetadata.prepare. For now I copy the same logic to `EmbeddingModelRunner.prepare_input_tensors()` except for SamplingMetadata. I apologize that I don't have the bandwidth to design this further.

robertgshaw2-neuralmagic commented:
Comments above.

Big items:
- The CI is red.

Small items:
- [ ] Revert `LLM.generate` interface change. The interface for `LLM.generate` changed (because we formerly had `RequestOutput` and now we expose `CompletionRequestOutput`. This is a breaking change. Its suboptimal, but I think we should have `RequestOutput` for `LLM.generate()` and `EmbeddingRequestOutput` for `LLM.encode()`. This will avoid a breaking change.
- [ ] Expand Test Cases. I like the test framework. Can you expand it to make sure there is some batching that gets triggered?

Questions:
- Have you run this though MTEB or anything that can do a holistic of correctness?
- Have you run any benchmarking to see if these changes impact performance of the generative models at all?

In terms of follow ups:
- Are you planning to continue on the roadmap / refactor that I laid out?

CatherineSue commented on vllm/model_executor/models/llama.py:
```diff
@@ -293,6 +293,44 @@ def forward(
         hidden_states, _ = self.norm(hidden_states, residual)
         return hidden_states
 
+    def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
```
`LlamaEmbeddingModel` uses `self.model = LlamaModel(**)`, how does it work if we directly put this in LlamaEmbeddingModel? The parameters are still in LlamaModel, right?

Commit:
Apply suggestions from code review

Co-authored-by: Roger Wang <136131678+ywang96@users.noreply.github.com>

CatherineSue committed:
Resolve comments

- Fix tests
- Add PoolingTensor in pooling_metadata.py
- Use PoolingMetadata instead of AttentionMetadata in pooler
- Revert LLM.generate() to return RequestOutput
- Add batch tests for embedding in test_openai_server.py
- Fix CI and remove --embedding-mode in arg_utils

CatherineSue commented on vllm/engine/async_llm_engine.py:
```diff
@@ -574,7 +582,7 @@ async def generate(
         lora_request: Optional[LoRARequest] = None,
         multi_modal_data: Optional[MultiModalData] = None
-    ) -> AsyncIterator[RequestOutput]:
+    ) -> AsyncIterator[CompletionRequestOutput]:
```
Done.

CatherineSue commented on vllm/core/embedding_model_block_manager.py:
```diff
@@ -0,0 +1,84 @@
+from vllm.sequence import Sequence, SequenceGroup
+
+
+class EmbeddingModelBlockSpaceManager(BlockSpaceManager):
```
The interfaces here are actively called in `Scheduler.schedule`. If we raise ValueError it would interrupt the schedule function.  I think it would make more sense to separate the schedule logic of embedding as it is pretty straightforward without block tables.

CatherineSue commented on tests/models/test_llama_embedding.py:
```diff
@@ -0,0 +1,47 @@
+
+@pytest.mark.parametrize("model", MODELS)
+@pytest.mark.parametrize("dtype", ["bfloat16"])
+# @pytest.mark.skip(
```
Removed.

CatherineSue commented on vllm/config.py:
```diff
@@ -177,6 +181,11 @@ def _verify_cuda_graph(self) -> None:
         self.max_context_len_to_capture = min(self.max_context_len_to_capture,
                                               self.max_model_len)
 
+    def _check_embedding_mode(self) -> bool:
```
Done

CatherineSue commented on vllm/config.py:
```diff
@@ -569,13 +579,17 @@ def __init__(
             if enable_chunked_prefill:
                 # For chunked prefill, choose the well-tuned batch size.
                 self.max_num_batched_tokens = 768
+            elif embedding_mode:
```
Done

CatherineSue commented:
Resolved most comments.

> The CI is red. Looks like you are missing a dependency. Add it to requirements-dev.txt
Added the dep. Waiting for CI.

- [x] Revert LLM.generate interface change. 
> The interface for LLM.generate changed (because we formerly had RequestOutput and now we expose CompletionRequestOutput. This is a breaking change. Its suboptimal, but I think we should have RequestOutput for LLM.generate() and EmbeddingRequestOutput for LLM.encode(). This will avoid a breaking change.

I reverted the API. Does this mean I need to revert CompletionRequestOutput back to RequestOutput?

Questions:

> Have you run this though MTEB or anything that can do a holistic of correctness?
I tried but `LLM.encode()` interface is not consistent with `SentenceTransformer`. And it would raise errors in MTEB.
We have performed correctness testing internally with our science team for their use case.

Have you run any benchmarking to see if these changes impact performance of the generative models at all?
> No, it is a valuable benchmark to try. I didn't have time to run regression test for the generative models. Is there any regression benchmark in CI? 

In terms of follow ups:

> Are you planning to continue on the roadmap / refactor that I laid out?
Unfortunately, I don't have bandwidth to continue the future tasks on the roadmap.

CatherineSue commented:
@ywang96 @robertgshaw2-neuralmagic I resolved most of the comments. Please see the summary above.

CatherineSue commented:
ModelTest keeps failing due to OOM. I tried to set `dtype` to `half` in `test_llama_embedding`. But it didn't seem to help. I ran it alone locally and the test passed.

<img width="904" alt="Screenshot 2024-04-24 at 11 47 18 PM" src="https://github.com/vllm-project/vllm/assets/8605658/bbc2b69b-9ebe-42c9-bad2-2ce9943fd364">

Is the model size too big?

CatherineSue committed:
Fix EntryPointsTest, ModelsTest and rebase

- Fix rebase errors
- Format files
- Suppress mypy type error in embedding_block_manager
- Add spec_decode_worker_metrics in PoolerOutput since it is required
in LLMEngine._get_stats

Fix EntryPoints Test
- It keeps running after introducing embedding_server in
test_openai_server.py. Try to set fixture scope to `module` to reduce
ray resource.

robertgshaw2-neuralmagic commented:
@CatherineSue shoot me a note when ready!

CatherineSue commented:
@robertgshaw2-neuralmagic It is ready. Not sure how to fix Entrypoints Test's ray resource issue or Models Test's OOM error. Suggestions appreciated.

robertgshaw2-neuralmagic commented:
@CatherineSue Ill take a look at getting the CI to pass

robertgshaw2-neuralmagic commented:
@CatherineSue we have release on Friday, so I think we should merge this after

CatherineSue commented:
@robertgshaw2-neuralmagic sounds good. Do you plan to take another look at this PR next week then?

robertgshaw2-neuralmagic commented:
I have reviewed in detail and feel good about it, just have been waiting for a clean CI build

WuNein commented:
@CatherineSue  https://github.com/CatherineSue/vllm/blob/1c508625a85449c83c8fc1f2f99d78e7035fcbb6/vllm/model_executor/layers/pooler.py#L13 
Last token is good, there are some using meaning pooling either.

WuNein commented:
Another question , this push will only solve llama's embedding problem.
https://github.com/vllm-project/vllm/blob/1c508625a85449c83c8fc1f2f99d78e7035fcbb6/vllm/model_executor/models/llama_embedding.py#L29

Support of other model, is a question?

CatherineSue commented:
@WuNein 
>"mean", "max", "cls", "weightedmean", "lasttoken" are all important.
>Another question , this push will only solve llama's embedding problem.

These are not within the scope of this PR. I believe there will be following PRs to cover these topics.

WuNein commented:
@CatherineSue For me, return the last hidden state is everything. Pool can be implement separately. You offer a solution that is more difficult to use on other model.

robertgshaw2-neuralmagic commented:
> @CatherineSue For me, return the last hidden state is everything. Pool can be implement separately. You offer a solution that is more difficult to use on other model.

We can expand upon this in a follow up PR

robertgshaw2-neuralmagic commented:
@CatherineSue I force pushed a commit that fixes the model test in the CI

robertgshaw2-neuralmagic commented:
Going to let the CI run now

CatherineSue commented:
@robertgshaw2-neuralmagic thanks!

robertgshaw2-neuralmagic commented:
> Another question , this push will only solve llama's embedding problem.
> 
> https://github.com/vllm-project/vllm/blob/1c508625a85449c83c8fc1f2f99d78e7035fcbb6/vllm/model_executor/models/llama_embedding.py#L29
> 
> Support of other model, is a question?

Adding other models in future PRs!

robertgshaw2-neuralmagic commented:
@CatherineSue Are you going to revert `CompletionRequestOutput` back to `RequestOutput`?

CatherineSue commented:
@robertgshaw2-neuralmagic I can do that now. I am not clear on the exact solution. Currently `CompletionRequestOutput` and `EmbeddingRequestOutput` are inherited from `RequestOutput` interface. To revert `CompletionRequestOutput` back, it would mean to remove the interface and maintain `RequestOutput` and `EmbeddingRequestOutput` separately, is that right?

CatherineSue committed:
Revert `CompletionRequestOutput` to `RequestOutput`

CatherineSue committed:
Update EmbeddingModelBlockSpaceManager interface

robertgshaw2-neuralmagic commented:
@CatherineSue ping me when ready

CatherineSue committed:
Move sentence-transformers to requirements-common.txt

- For AMD Tests

CatherineSue commented:
@robertgshaw2-neuralmagic Code change is done. Some new CIs are failing:
AMD Benchmark Test: Connection Refused.
Speculative decoding tests: CUDA OOM
Models Test: CUDA OOM

CatherineSue commented:
@robertgshaw2-neuralmagic I did another rebase. Looks like Speculative decoding and Models tests are still failing due to CUDA OOM. I can't seem to find a solution here. Can I get some suggestions from you? Thanks!

robertgshaw2-neuralmagic commented on tests/conftest.py:
```diff
@@ -162,6 +160,16 @@ def __init__(
                 model_name,
                 torch_dtype=torch_dtype,
             )
+        elif model_name in _EMBEDDING_MODELS:
```
@CatherineSue here is where you are running OOM

I pushed a fix for this to your branch, but looks like it got overridden

The model is getting loaded at fp32 here, so its consuming 7*4=28GB or RAM on a machine with 24GB of RAM

Load the model at FP16 here and you will be good to go

CatherineSue replied:
thanks!! applied. Running CI now.

CatherineSue replied:
Seems the test is still failing. 

>model-00001-of-00002.safetensors: 100% 9.94G/9.94G [00:55<00:00, 180MB/s]
  model-00002-of-00002.safetensors: 100% 4.28G/4.28G [00:23<00:00, 183MB/s]

I have changed the dtype to fp16.

CatherineSue replied:
Passed device=cpu in SentenceTransformer init. This test is passed in Models Test now.

CatherineSue committed:
Fix Models Test and update interface for embedding_block_manager

CatherineSue committed:
Rebase

CatherineSue commented:
Latest test on llama_embedding
<img width="941" alt="Screenshot 2024-05-08 at 11 32 02 PM" src="https://github.com/vllm-project/vllm/assets/8605658/bdc43157-d3e5-4c2b-8b8f-6b52682c370a">

CatherineSue committed:
Fix Models Test

CatherineSue commented:
@robertgshaw2-neuralmagic The current failing AMD Distributed Test is not caused by changes in this PR.

CatherineSue commented:
@robertgshaw2-neuralmagic Do you think the PR is ready for merging? Given that the vLLM is actively maintained, we're encountering more frequent merge conflicts. Since this PR is extensive, rebasing it has become quite time-consuming. Could we aim to merge it soon?

robertgshaw2-neuralmagic commented:
> @robertgshaw2-neuralmagic Do you think the PR is ready for merging? Given that the vLLM is actively maintained, we're encountering more frequent merge conflicts. Since this PR is extensive, rebasing it has become quite time-consuming. Could we aim to merge it soon?

Yes - going to do final review + merge.

Thank you so much for your efforts on this

robertgshaw2-neuralmagic commented:
Going to force push a few things to this branch @CatherineSue - please do not make any more changes

robertgshaw2-redhat committed:
Merge branch 'main' into embedding

Commit:
format

Commit:
added test_embedding

Commit:
added examples

Commit:
cleanup

Commit:
cleanup

Commit:
cleanup

Commit:
cleanup

Commit:
new line

Commit:
reducing changes

Commit:
simplify test changes

Commit:
simplify test changes

Commit:
simplify test changes

Commit:
simplify test changes

Commit:
style for setting up embedding mode in model_config

Commit:
nit on engineargs

Commit:
updated comment

Commit:
cleanup

Commit:
removed change from llama.py

Commit:
final review

Commit:
final review

Commit:
format

robertgshaw2-neuralmagic commented:
@CatherineSue 

Thank you so much for your efforts. This is ready to go. Just letting the CI run and then will merge.

CatherineSue commented:
@robertgshaw2-neuralmagic thank you so much for all the time and help on reviews and CIs!!

Commit:
Merge branch 'main' into embedding

robertgshaw2-redhat committed:
Update conftest.py

Opdoop commented:
@CatherineSue Thanks for the incredible works! Is there a speed comparison between vllm and hf/sentence-transformers in terms of inference speed?

K-Mistele commented:
Are there any updates to the project docs on how to actually use this feature?

linuxliker commented:
Are there any updates to the project docs on how to actually use this feature?

CatherineSue commented:
@linuxliker @K-Mistele this feature is ready to use. You can follow [examples/openai_embedding_client.py](https://github.com/vllm-project/vllm/blob/main/examples/openai_embedding_client.py) to use it.

yannickschuchmann commented:
Thanks for the effort @CatherineSue 🙏 
The openai example leaves me confused unfortunately:

There is no model name configured in the example. 
Does that mean when I use vllm with the mistral model ("mistralai/Mistral-7B-Instruct-v0.2") the embeddings endpoint is automatically enabled as well? 
Or do I have to run two vllm instances? one "intfloat/e5-mistral-7b-instruct" for embeddings and one "mistralai/Mistral-7B-Instruct-v0.2" for completions endpoint? 

I flew over the code, but unfortunately couldn't figure out how to set an embedding and a regular model at the same time.

Thanks in advance :)

CatherineSue commented:
@yannickschuchmann thanks for trying this out. You would need to run two vLLM instances. Running one vLLM instance with two models is not supported for now.

nbroad1881 commented:
If a model's tokenizer has `add_bos_token=True` and `add_eos_token=True`, will vLLM handle this, or should the special tokens be inserted into the text prompt?
